server/src/main/java/org/apache/uniffle/server/storage/MultiStorageManager.java

```diff
@@ -65,39 +95,52 @@ public Storage selectStorage(ShuffleDataReadEvent event) {
 
   @Override
   public void updateWriteMetrics(ShuffleDataFlushEvent event, long writeTime) {
-    selectStorageManager(event).updateWriteMetrics(event, writeTime);
+    throw new UnsupportedOperationException();
   }
 
   @Override
   public boolean write(Storage storage, ShuffleWriteHandler handler, ShuffleDataFlushEvent event) {
     StorageManager storageManager = selectStorageManager(event);
-    if (storageManager == coldStorageManager && event.getRetryTimes() > fallBackTimes) {
+    if (event.getRetryTimes() > 0) {
       try {
-        CreateShuffleWriteHandlerRequest request = storage.getCreateWriterHandlerRequest(
-            event.getAppId(),
-            event.getShuffleId(),
-            event.getStartPartition());
-        if (request == null) {
-          return false;
+        StorageManager newStorageManager = storageManagerFallbackStrategy.tryFallback(
+                storageManager, event, warmStorageManager, coldStorageManager);
+        if (newStorageManager != storageManager) {
+          storageManager = newStorageManager;
+          storageManagerCache.put(event, storageManager);
+          CreateShuffleWriteHandlerRequest request = storage.getCreateWriterHandlerRequest(
+              event.getAppId(), event.getShuffleId(), event.getStartPartition());
+          storage = storageManager.selectStorage(event);
+          handler = storage.getOrCreateWriteHandler(request);
```

Comment:
After reviewing this part, I think this is a bit strange that change the external objects reference of `storage` and `handler`. This will make the external invoker like `ShuffleFlushManager` confused.

@jerqi @xianjingfeng

Reply:
> After reviewing this part, I think this is a bit strange that change the external objects reference of `storage` and `handler`. This will make the external invoker like `ShuffleFlushManager` confused.
> 
> @jerqi @xianjingfeng

We encapsulate the behavior. `ShuffleFlushManger` don't need care.

Reply:
If you have better solution, you can propose it.

Reply:
Currently I have no ideas. I just want to use this fallback strategy and review this part.

Reply:
> Currently I have no ideas. I just want to use this fallback strategy and review this part.

Flush process worry me a lot. I'm not satisfied about the code quality. But I don't have idea how to improve them. Maybe @LuciferYang @advancedxy can help us.

Reply:
I have two directions to improve this part of MultiStorageManager

1. Remove the `storageManagerCache` , it looks unused and bring some problems. One possible problem I have seen is that, when one event enters into pending queue due to storage cannot write, it will not have a chance to get a new fallback strategy due to cache.
2. Avoid changing the external reference object in `write` method. If we want to fallback write to other storage. We could use the invoking sequence like. (selectStorage -> write)  if failed and then (selectStorage -> write), instead of selectStorage -> write (failed) -> write(choose other storage in this method). That means we should change the fallback strategy invoked from `write` method to `selectStorage` method.

Reply:
> I have two directions to improve this part of MultiStorageManager
> 
> 1. Remove the `storageManagerCache` , it looks unused and bring some problems. One possible problem I have seen is that, when one event enters into pending queue due to storage cannot write, it will not have a chance to get a new fallback strategy due to cache.
> 2. Avoid changing the external reference object in `write` method. If we want to fallback write to other storage. We could use the invoking sequence like. (selectStorage -> write)  if failed and then (selectStorage -> write), instead of selectStorage -> write (failed) -> write(choose other storage in this method). That means we should change the fallback strategy invoked from `write` method to `selectStorage` method.

Agree